### PR TITLE
feat: Updated repr for ParametrizedHamiltonian

### PR DESCRIPTION
**Pull request by anushkrishnav**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
The context of this change is to provide an informative string representation for the ParametrizedHamiltonian class in PennyLane.

**Description of the Change:**
The change involves updating the __repr__ method of the ParametrizedHamiltonian class to generate a string representation that closely follows the example provided. This includes incorporating the parametrized functions and their corresponding indices in the expression, using the appropriate brackets.

**Benefits:**

The updated string representation will provide users with a more informative view of the ParametrizedHamiltonian object. It will clearly show the terms of the Hamiltonian, including any parametrized functions, their corresponding parameter indices, and the time variable.

**Possible Drawbacks:**
One potential drawback is that the string representation may become more complex for Hamiltonians with a large number of terms or deeply nested expressions. This could make it harder to read and understand for very complex Hamiltonians.

**Related GitHub Issues:**
#4088

### CODE REVIEW

File: pennylane/pulse/parametrized_hamiltonian.py
Comment (Qottmann) [inline pennylane/pulse/parametrized_hamiltonian.py:248]: I think we should separate iterating over `.._fixed` and `.._parametrized` as the value of `i` is offset by the number of fixed terms

File: pennylane/pulse/parametrized_hamiltonian.py
Comment (Qottmann) [inline pennylane/pulse/parametrized_hamiltonian.py:250]: We have to be a bit more careful here because also class objects are `callable`. The following example would raise an error:

```python
import pennylane as qml

def f1(p, t):
    return p[0] * jnp.sin(p[1] * t)

def f2(p, t):
    return p * t

class f3:
    def __init__(self, hyper_param):
        self.hyper = hyper_param
    def __call__(self, p, t):
        return self.hyper * p * t

coeffs = [2., f1, f2, f3(0.5)]
observables =  [qml.PauliX(0), qml.PauliY(0), qml.PauliZ(0), qml.PauliX(0)]
H = qml.dot(coeffs, observables)

print(H)
```

```pycon
AttributeError: 'f3' object has no attribute '__name__'
```

**Review by Qottmann — CHANGES_REQUESTED** [review]
Hello @anushkrishnav thank you for your contribution!
This already looks like it is on a good way :-)

There are two issues with the current code:
1. the index i is not displaying the correct number
2. callable class objects raise an error because they dont have a `__name__` attribute

Additionally, it seems there is a failing test in [tests/pulse/test_parametrized_hamiltonian.py::TestInitialization::test__repr__](https://github.com/PennyLaneAI/pennylane/blob/master/tests/pulse/test_parametrized_hamiltonian.py#L125)
You can update this test, and I would suggest adding another test to check the new feature behaves as you intend it to. In order to check the test is passing, you can run `python -m pytest tests/pulse/test_parametrized_hamiltonian.py` (or any other relevant file) in the terminal before pushing.

The following is optional:
There is a special subclass of `ParametrizedHamiltonian`, [`HardwareHamiltonian`](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/pulse/hardware_hamiltonian.py). Let me tell you what the situation is and how it is handled in `HardwareHamiltonian`.
For transmon or rydberg systems, the driving term of the Hamiltonian typically has the form
$H_\text{drive} = \Omega(p, t) (\cos(\nu t) X - \sin(\nu t) Y)$.
When implementing this in pennylane right now, this yields two separate terms $\Omega(p, t) \cos(\nu t) X$ and $-\Omega(p, t) \sin(\nu t) Y$, sharing the same parameters $(p, \nu)$. In order to avoid having to pass the parameters twice to the function, there is a `_reorder_parameters` function that reorders the parameters that are passed (see https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/pulse/hardware_hamiltonian.py#L237 for rydberg systems and https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/pulse/transmon.py#L502 for transmons.
That means, the string representation of `HardwareHamiltonian` (which inhertis from `ParametrizedHamiltonian`) would have to be overwritten with a slightly modified function that takes that reordering into account.

It is slightly more tricky, so optional.

**Comment by anushkrishnav** [comment]
Hey thanks for the detailed review @Qottmann  I will start working on the requested changes, Will update the tests as well.

File: pennylane/pulse/parametrized_hamiltonian.py
Comment (anushkrishnav) [reply pennylane/pulse/parametrized_hamiltonian.py:250]: I think a simple ```hasattr(coeff, "__name__")```
Should fix this problem

File: pennylane/pulse/parametrized_hamiltonian.py
Comment (anushkrishnav) [reply pennylane/pulse/parametrized_hamiltonian.py:248]: Yes WIll do it , I will split them into 2 different iteration like this 
```
        for coeff, op in zip(self.coeffs_fixed, self.ops_fixed):
            term = f"{coeff} * {op}"
            terms.append(term)

        for i, (coeff, op) in enumerate(zip(self.coeffs_parametrized, self.ops_parametrized)):
            if callable(coeff) and hasattr(coeff, "__name__")
                term = f"{coeff.__name__}(params_{i}, t) * {op}"
            else:
                term = f"{coeff} * {op}"
            terms.append(term)
            ```

**Comment by anushkrishnav** [comment]
I have made the requested updates @Qottmann 
Kindly take a look, 
Thank you.

**Comment by anushkrishnav** [comment]
Yeh, the optional one sounds interesting do share more info I will work on it as an individual PR, I am currently looking at the files you share to get a better grasp of whats actually happening and the reodering

**Review by Qottmann — COMMENTED** [review]
Getting there, great work @anushkrishnav 💪 

Note that lines 253 and 255 are currently not covered by the tests (this is what codecov checks and complains about in the test suite below). In order to fix that, you need to add tests that cover these cases. I.e. test the string representation with a `ParametrizedHamiltonian` that contains a callable class object with a `__name__`, and another test with a callable class object with no `__name__` and assert that it outputs what you expect.

Some optional refinement (not needed for unitary hack, but I think it would make the PR overall nicer): Currently everything is returned in one line. It would be great to output terms in separate lines like is done in `qml.Hamiltonian`, see [ops/qubit/hamiltonian.py#LL490](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/ops/qubit/hamiltonian.py#LL490C7-L490C7) for inspiration. See example here:
image


Since we are changing the string representation, we also need to make sure the outputs that were hardcoded in the docs are updated. This concerns, as far as I can see, the doc strings in:
- the module page https://docs.pennylane.ai/en/latest/code/qml_pulse.html
- `ParmetrizedHamiltonian` https://docs.pennylane.ai/en/latest/code/api/pennylane.pulse.ParametrizedHamiltonian.html
(Cant guarantee I didnt miss another public doc string in the pulse module)
I.e. this would need to be updated in the current docs:
image
(You can render docs locally by running `make docs` to check them, or click on "details" in the readthedocs item in the test suite:)
image


If you want to do the case of `HardwareHamiltonian` in a separate PR, already this PR should overwrite the `__repr__` to do the default because it currently outputs the wrong string. I will write a separate message to provide more details to continue there :)

**Comment by Qottmann** [comment]
Re `HardwareHamiltonian`:
This is a special class to look at Hamiltonians that can be executed on a real device. We are currently supporting transmon (superconducting qubits) and rydberg (neutral atom) systems. Let us look at the case for transmons (rydberg is very similar). The Hamiltonian has the form

$H = H_\text{int} + \sum_i \Omega_i(t) (\phi_i + \cos(\nu_i t) X_i - \sin(\phi_i + \nu_i t) Y_i)$

where $H_\text{int}$ is some fixed interaction Hamiltonian. There are two complications:
a) All three of $\phi_i$, $Omega_i(t)$ and $\nu_i$ can be callable, so we have two functions, each expecting a parameter to be passed to it (this is a problem because `ParametrizedHamiltonian` always expects the scalar callables to have the signature `f(p, t)`)
b) Two different terms, $\Omega_i(t) \cos(\phi_i + \nu_i t) X_i$, and $\Omega_i(t) \sin(\phi_i + \nu_i t) Y_i$ share the parameters in `\Omega_i` and `\nu_i`.
If you were to implement this Hamiltonian by hand in PennyLane, you need some workarounds. I did this in [this demo](https://pennylane.ai/qml/demos/tutorial_pulse_programming101.html):
image

To allow users to be flexible in whether they want $\phi_i$, $\nu_i$ or $\Omega_i$ to be callable or not, and not have to pass the parameters twice, we introduced the `reorder_fn` under the hood in `HardwareHamiltonian`. Internally, it uses a special callable class object `AmplitudeAndPhaseAndFreq` that checks which are callable and which one is not, and accordingly uses the right function as its call. On the other hand, the reorder function checks which of the three are callable, and passes twice, depending on how many are callable, the next 1, 2 or 3 parameters.

The string representation would basically have to follow the same logic as in https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/pulse/transmon.py#L502 i.e. check whether it is an instance of this special callable class, and if yes count the number of callables and accordingly increase the index that is being displayed.

So ideally the string representation for the following Hamiltonian would look like this
```
qubit_freq = [1., 1.]
connections = [(0, 1)]
coupling = 0.01831
wires = [0, 1]

def f1(p, t):
    return p * t

def f2(p, t):
    return jnp.sin(p * t)

H = qml.pulse.transmon_interaction(qubit_freq, connections, coupling, wires)
H += qml.pulse.transmon_drive(f1, f1, 0.5, wires=[0])
H += qml.pulse.transmon_drive(f1, 0.2, f2, wires=[1])

>>> print(H)
(6.283185307179586*(((0.5*(PauliX(wires=[0]))) + ((-0-0.5j)*(PauliY(wires=[0])))) @ ((0.5*(PauliX(wires=[0]))) + (0.5j*(PauliY(wires=[0]))))))
+ (6.283185307179586*(((0.5*(PauliX(wires=[1]))) + ((-0-0.5j)*(PauliY(wires=[1])))) @ ((0.5*(PauliX(wires=[1]))) + (0.5j*(PauliY(wires=[1]))))))
+ (0.11504512297445822*((((0.5*(PauliX(wires=[0]))) + ((-0-0.5j)*(PauliY(wires=[0])))) @ ((0.5*(PauliX(wires=[1])))
+ (6.283185307179586*(((0.5*(PauliX(wires=[1]))) + ((-0-0.5j)*(PauliY(wires=[1])))) @ ((0.5*(PauliX(wires=[1]))) + (0.5j*(PauliY(wires=[1])))))) + (0.06283185307179587*((((0.5*(PauliX(wires=[0]))) + ((-0-0.5j)*(PauliY(wires=[0])))) @ ((0.5*(PauliX(wires=[1]))) + (0.5j*(PauliY(wires=[1]))))) + (((0.5*(PauliX(wires=[1]))) + ((-0-0.5j)*(PauliY(wires=[1])))) @ ((0.5*(PauliX(wires=[0]))) + (0.5j*(PauliY(wires=[0])))))))
+ (f1(params_0, t) cos(f1(params_1, t) + 0.5 * t) * (PauliX(wires=[0]))) + (f1(params_0, t) cos(f1(params_1, t) + 0.5 * t) * *(-1*(PauliY(wires=[0]))))
+ (f1(params_2, t) cos(0.2 + f2(params_3, t) * t) * (PauliX(wires=[1]))) + (f1(params_2, t) cos(0.2 + f2(params_3, t) * t) * (-1*(PauliY(wires=[1])))
```
This is not super nice looking because internally the annihilation and creation operators are explicitly constructed as $a = 1/2(X + i Y)$, I'm afraid it would go beyond the scope of this PR to fix that as well.
Some further refinement could be to write the $6.283185307179586 = 2 \pi$ factors explicitly.

File: tests/pulse/test_parametrized_hamiltonian.py
Comment (Qottmann) [inline tests/pulse/test_parametrized_hamiltonian.py]: Since this is not used anywhere else, I would move this inside the test

File: tests/pulse/test_parametrized_hamiltonian.py
Comment (Qottmann) [inline tests/pulse/test_parametrized_hamiltonian.py:146]: Sorry for being pedantic, but can you please separate this out into a second test. Something like `test_repr_with_class_objects`

File: pennylane/pulse/rydberg.py
Comment (Qottmann) [inline pennylane/pulse/rydberg.py:215]: These are actually unwanted because they are not correct. We should fix the `__repr__` of `HardwareHamiltonian`

**Review by Qottmann — CHANGES_REQUESTED** [review]
Thanks for the update @anushkrishnav 🚀 
The line breaks make it very clean, nice!

There are missing outputs in the docstring of `ParametrizedHamiltonian`

We cant merge the PR as long as the transmon and rydberg Hamiltonians are producing erroneous outputs. Since [HardwareHamiltonian](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/pulse/hardware_hamiltonian.py#L271) inherits from `ParametrizedHamiltonian`, it is using the same `__repr__` method. But in that case, we want the default back because the logic that you added does not apply to those Hamiltonians as described and discussed in the potential follow-up PR.

File: tests/pulse/test_parametrized_hamiltonian.py
Comment (anushkrishnav) [reply tests/pulse/test_parametrized_hamiltonian.py:146]: Certainly

File: pennylane/pulse/rydberg.py
Comment (Qottmann) [reply pennylane/pulse/rydberg.py]: Yes, we cant assume the follow up PR will be done and merged, so this should work independently correctly and as expected. Since the case of hardware hamiltonian is more complex, I would revert to the default behavior here (i.e. change the `__repr__` and the outputs in the docs accordingly)

**Comment by Qottmann** [comment]
Note also that currently pylint and black are failing!
See https://docs.pennylane.ai/en/stable/development/guide/pullrequests.html

**Comment by anushkrishnav** [comment]
Yeh I saw that pylint my next commit will fix that

File: pennylane/pulse/rydberg.py
Comment (Qottmann) [reply pennylane/pulse/rydberg.py]: yep! Only a temporary fix (assuming we'll continue with providing a nice `__repr__` in a follow-up :)

Comment (anushkrishnav) [reply pennylane/pulse/rydberg.py]: I will do this, I read your explanation on hardware, thanks for going in detail about it learning a lot about something I didn't know much of , I am learning more about the system for the past few days . Thank you

**Review by Qottmann — APPROVED** [review]
Thanks for the update @anushkrishnav 💪 

Looks almost good to go!
There are some missing output changes in the docstrings!

In the `rydberg_interaction` docstring:
image
in `rydberg_drive`
image
in `transmon_interaction`
image

We require two approving reviews before merging, another colleague will do a code-review in the following days. There may be things that I have missed and further changes necessary for them, but overall it is looking good to me 👍

File: pennylane/pulse/rydberg.py
Comment (lillian542) [inline pennylane/pulse/rydberg.py:203]: I believe both of these return the `HardwareHamiltonian: terms=X" representation now

**Review by lillian542 — COMMENTED** [review]
This is looking great, thanks! 🎉 

I only have a couple of very minor comments:

There is an example in `pennylane/ops/functions/dot.py` that still uses the old representation in the docstring. I also see a few docstring examples in the `transmon.py` and `rydberg.py` files that show `ParametrizedHamiltonian` instead of `HardwareHamiltonian`.

Don't forget to add a change-log entry and include yourself in the list of contributors!

File: pennylane/pulse/rydberg.py
Comment (anushkrishnav) [reply pennylane/pulse/rydberg.py:203]: Yes, sorry on the delay will push the updated docstring soon and add myself in changelog

**Review by lillian542 — APPROVED** [review]
Nice! Thanks @anushkrishnav 🎉

**Comment by anushkrishnav** [comment]
> Nice! Thanks @anushkrishnav tada

Thank you ! I will look into Hardware soon , I think I will be able to get it done too

**Pull request merged**
